mr/src/main/java/org/opensearch/hadoop/rest/commonshttp/CommonsHttpTransport.java

```diff
@@ -712,12 +793,95 @@ public Object run() throws Exception {
+            for (String value : param.getValue()) {
+                result.add(SdkHttpUtils.urlEncode(param.getKey()) + '=' + SdkHttpUtils.urlEncode(value));
+            }
+        }
+
+        return '?' + AMPERSAND_JOINER.join(result.build());
+    }
+
+    private void awsSigV4SignRequest(Request request, HttpMethod http, byte[] bodyBytes)
+            throws UnsupportedEncodingException {
+        String awsRegion = settings.getAwsSigV4Region();
+        String awsServiceName = settings.getAwsSigV4ServiceName();
+
+        Region signingRegion = Region.of(awsRegion);
+
+        final AwsCredentialsProvider credentials = DefaultCredentialsProvider.create();
+
+        Aws4SignerParams signerParams = Aws4SignerParams.builder()
+                .awsCredentials(credentials.resolveCredentials())
+                .signingName(awsServiceName)
+                .signingRegion(signingRegion)
+                .build();
+
+        SdkHttpFullRequest.Builder req = SdkHttpFullRequest.builder()
+                .method(SdkHttpMethod.fromValue(request.method().name()));
+
+        StringBuilder url = new StringBuilder();
+        url.append(httpInfo);
+        String path = request.path().toString();
+        if (!path.startsWith("/")) {
+            url.append('/');
+        }
+        url.append(path);
+
+        Splitter queryStringSplitter = Splitter.on('&').trimResults().omitEmptyStrings();
+        final Iterable<String> rawParams = request.params() != null ? queryStringSplitter.split(request.params())
+                : Collections.emptyList();
+
+        final ImmutableListMultimap.Builder<String, String> queryParams = ImmutableListMultimap.builder();
+
+        for (String rawParam : rawParams) {
+            if (!Strings.isNullOrEmpty(rawParam)) {
+                final String pair = URLDecoder.decode(rawParam, StandardCharsets.UTF_8.name());
+                final int index = pair.indexOf('=');
+                if (index > 0) {
+                    final String key = pair.substring(0, index);
+                    final String value = pair.substring(index + 1);
+                    queryParams.put(key, value);
+                } else {
+                    queryParams.put(pair, "");
+                }
+            }
+        }
+
+        String params = queryParamsString(queryParams.build());
+
+        url.append(params);
+
+        try {
+            req.uri(new java.net.URI(url.toString()));
+        } catch (URISyntaxException e) {
+            throw new IllegalArgumentException("Invalid request URI: " + request.uri().toString());
+        }
+
+        if (request.body() != null) {
+            req.contentStreamProvider(() -> new ByteArrayInputStream(bodyBytes));
+        }
+        req.putHeader("x-amz-content-sha256", "required");
+
+        SdkHttpFullRequest signedRequest = Aws4Signer.create().sign(req.build(), signerParams);
+
+        final ImmutableMap.Builder<String, String> signerHeaders = ImmutableMap.builder();
```

Comment:
May be have this method return a signed request and let the caller logic take care of adding headers to `http`?